swift/Lottie.swift

```diff
@@ -127,27 +128,61 @@ public class Lottie {
     }
 
     /// Prepares the canvas for rendering by setting the frame, adjusting the picture size, and optionally applying a crop.
-    private func prepareCanvasForRendering(frameIndex index: Int, size: CGSize, crop: CGRect?) {
+    private func prepareCanvasForRendering(frameIndex index: Int, size: CGSize, crop: CGRect?) throws {
         tvg_animation_set_frame(animation, Float(index))
 
         let picture = tvg_animation_get_picture(animation)
         tvg_picture_set_size(picture, Float(size.width), Float(size.height))
 
         if let crop {
-            let cropShape = tvg_shape_new()
+            try applyCropping(crop, relativeTo: size)
+        }
+
+        tvg_canvas_update_paint(canvas, picture)
+    }
+
+    /// Applies cropping to the Lottie animation by resizing and translating the picture to fit within a specified cropping rectangle, relative to a given size.
+    /// Note: The function ensures that the cropped area is scaled and positioned correctly to fit within the specified size, maintaining the aspect ratio of the cropped area.
+    private func applyCropping(_ crop: CGRect, relativeTo size: CGSize) throws {
+        guard crop.width <= size.width, crop.height <= size.height else {
+            throw LottieError.croppingRectangleOutsideOfFrameBounds
+        }
+
+        let picture = tvg_animation_get_picture(animation)
+
+        let cropShape = tvg_shape_new()
```

Comment:
from line 153 to 184, i guess it's doing:
1. crop the picture
2. scale the crop to the original size
3. move x and y
4. then mask it to get the final result

Somehow I think we may be able to simplify the steps, but we can leave it. Unless there is a performance difference, which I believe is unlikely. 

But maybe adding some comments here can help

Reply:
Good call, I'll add some comments as it may be a bit confusing. 

I'll see if I can simplify the logic, because yeah having two cropping steps does seem a bit redundant.

Reply:
Eureka! I've simplified it. 

Now we just perform scaling, translation and then a single crop. Much easier!